common/archiver/s3store/historyArchiver.go

```diff
@@ -0,0 +1,333 @@
+			return archiver.ErrHistoryMutated
+		}
+
+		historyBatches = append(historyBatches, historyBlob.Body...)
+	}
+
+	encodedHistoryBatches, err := encode(historyBatches)
+	if err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
+		return err
+	}
+
+	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion)
+
+	if err := upload(ctx, h.s3cli, URI, key, encodedHistoryBatches); err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (h *historyArchiver) Get(
+	ctx context.Context,
+	URI archiver.URI,
+	request *archiver.GetHistoryRequest,
+) (*archiver.GetHistoryResponse, error) {
+	if err := h.ValidateURI(URI); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidURI.Error()}
+	}
+
+	if err := archiver.ValidateGetRequest(request); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidGetHistoryRequest.Error()}
+	}
+
+	var err error
+	var token *getHistoryToken
+	if request.NextPageToken != nil {
+		token, err = deserializeGetHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: archiver.ErrNextPageTokenCorrupted.Error()}
+		}
+	} else if request.CloseFailoverVersion != nil {
+		token = &getHistoryToken{
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+			NextBatchIdx:         0,
+		}
+	} else {
+		highestVersion, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *highestVersion,
+			NextBatchIdx:         0,
+		}
+	}
+
+	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
+	encodedHistoryBatches, err := download(ctx, h.s3cli, URI, key)
+	if err != nil {
+		return nil, err
+	}
+	historyBatches, err := decodeHistoryBatches(encodedHistoryBatches)
+	if err != nil {
+		return nil, &shared.InternalServiceError{Message: err.Error()}
+	}
+	historyBatches = historyBatches[token.NextBatchIdx:]
+
+	response := &archiver.GetHistoryResponse{}
+	numOfEvents := 0
+	numOfBatches := 0
+	for _, batch := range historyBatches {
+		response.HistoryBatches = append(response.HistoryBatches, batch)
+		numOfBatches++
+		numOfEvents += len(batch.Events)
+		if numOfEvents >= request.PageSize {
+			break
+		}
+	}
+
+	if numOfBatches < len(historyBatches) {
+		token.NextBatchIdx += numOfBatches
+		nextToken, err := serializeToken(token)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		response.NextPageToken = nextToken
+	}
+
+	return response, nil
+}
+
+func (h *historyArchiver) ValidateURI(URI archiver.URI) error {
```

Comment:
Is it possible to also add a test for if the bucket exists in S3 to the function? Same for the `ValidateURI` function in visibility archiver.

Reply:
This would mean an extra round trip to s3 for each archive call. The error will be caught later on if misconfigured. What do you think?

Reply:
Currently, once a history/visibility URI is set for a domain it can not be updated later, because otherwise data can be archived to multiple locations and we don't be able to retrieve it upon request. That's why it's important when setting the URI, we need to get it right at the first time. 

Having the existence check help catch some typos when setting URI. I am ok one extra call as it should be cheap because it's not uploading/downloading any data. Thoughts?

Reply:
Agreed, didn't think about that case. However it might not be necessary to use ValidateURI in the Archive() and Get() methods

Reply:
I agree. Maybe creating a util function to validate only the format? Archive() and Get() only call that util function while ValidateURI() will do an additional existence check?